output/cloud/expv2/output.go

```diff
@@ -0,0 +1,266 @@
+		Type:       mtype,
+		TimeSeries: as.MapAsProto(o.referenceID),
+	}
+}
+
+type aggregatedSamples struct {
+	Samples map[metrics.TimeSeries][]*metrics.Sample
+}
+
+func (as *aggregatedSamples) AddSample(s *metrics.Sample) {
+	tss, ok := as.Samples[s.TimeSeries]
+	if !ok {
+		// TODO: optimize the slice allocation
+		// A simple 1st step: Reuse the last seen len?
+		as.Samples[s.TimeSeries] = []*metrics.Sample{s}
+		return
+	}
+	as.Samples[s.TimeSeries] = append(tss, s)
+}
+
+func (as *aggregatedSamples) Clean() {
+	// TODO: evaluate if it makes sense
+	// to keep the most frequent used keys
+
+	// the compiler optimizes this
+	for k := range as.Samples {
+		delete(as.Samples, k)
+	}
+}
+
+func (as *aggregatedSamples) MapAsProto(refID string) []*pbcloud.TimeSeries {
+	if len(as.Samples) < 1 {
+		return nil
+	}
+	pbseries := make([]*pbcloud.TimeSeries, 0, len(as.Samples))
+	for ts, samples := range as.Samples {
+		pb := pbcloud.TimeSeries{}
+		// TODO: optimize removing Map
+		// and using https://github.com/grafana/k6/issues/2764
+		pb.Labels = make([]*pbcloud.Label, 0, ((*atlas.Node)(ts.Tags)).Len())
+		pb.Labels = append(pb.Labels, &pbcloud.Label{Name: "__name__", Value: ts.Metric.Name})
+		pb.Labels = append(pb.Labels, &pbcloud.Label{Name: "test_run_id", Value: refID})
+		for ktag, vtag := range ts.Tags.Map() {
+			pb.Labels = append(pb.Labels, &pbcloud.Label{Name: ktag, Value: vtag})
+		}
+
+		switch ts.Metric.Type {
+		case metrics.Counter:
+			counterSamples := &pbcloud.CounterSamples{}
+			for _, counterSample := range samples {
+				counterSamples.Values = append(counterSamples.Values, &pbcloud.CounterValue{
+					Time:  timestamppb.New(counterSample.Time),
+					Value: counterSample.Value,
+				})
+			}
+			pb.Samples = &pbcloud.TimeSeries_CounterSamples{
+				CounterSamples: counterSamples,
+			}
+		case metrics.Gauge:
+			gaugeSamples := &pbcloud.GaugeSamples{}
+			for _, gaugeSample := range samples {
+				gaugeSamples.Values = append(gaugeSamples.Values, &pbcloud.GaugeValue{
+					Time:  timestamppb.New(gaugeSample.Time),
+					Last:  gaugeSample.Value,
+					Min:   gaugeSample.Value,
+					Max:   gaugeSample.Value,
+					Avg:   gaugeSample.Value,
+					Count: 1,
+				})
+			}
+			pb.Samples = &pbcloud.TimeSeries_GaugeSamples{
+				GaugeSamples: gaugeSamples,
+			}
+		case metrics.Rate:
+			rateSamples := &pbcloud.RateSamples{}
+			for _, rateSample := range samples {
+				nonzero := uint32(0)
+				if rateSample.Value != 0 {
+					nonzero = 1
+				}
+				rateSamples.Values = append(rateSamples.Values, &pbcloud.RateValue{
+					Time:         timestamppb.New(rateSample.Time),
+					NonzeroCount: nonzero,
+					TotalCount:   1,
+				})
+			}
+			pb.Samples = &pbcloud.TimeSeries_RateSamples{
+				RateSamples: rateSamples,
+			}
+		case metrics.Trend:
+			trendSamples := &pbcloud.TrendHdrSamples{}
+			for _, trendSample := range samples {
+				hdrValue := histogramAsProto(
+					newHistogram([]float64{trendSample.Value}),
```

Comment:
Something is wonky here, we call `newHistogram()` with only a single sample value :confused: And we don't call `addToBucket()` anywhere.